bindings/pyroot_experimental/PyROOT/src/DeclareCallablePyz.cxx

```diff
@@ -0,0 +1,663 @@
+      return NULL;
+   }
+
+   err = gInterpreter->Declare(code_cstr);
+   if (!err) {
+      PyErr_SetString(PyExc_RuntimeError,
+              ("Failed to compile C++ wrapper: Compilation error from following wrapper code.\n" + code.str()).c_str());
+      return NULL;
+   }
+
+   // Pass through Python callable
+   return pyfunc;
+}
+
+
+// Call method of class used as decorator to create C++ wrapper using numba
+// The call method creates the C++ wrapper class for the Python callable and
+// passes through the actual callable.
+PyObject* NumbaCallableImpl_call(PyObject * /*self*/, PyObject *args)
+{
+   // Parse arguments with Python callable and get attributes with types
+   if (!PyTuple_Check(args)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to parse arguments: Invalid tuple.");
+      return NULL;
+   }
+
+   if (!(PyTuple_Size(args) == 2)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to parse arguments: Expect exactly one argument (Python callable).");
+      return NULL;
+   }
+
+   if (!(PyTuple_Size(args) == 2)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to parse arguments: Expect exactly one argument (Python callable).");
+      return NULL;
+   }
+
+   auto instance = PyTuple_GetItem(args, 0);
+   auto pyfunc = PyTuple_GetItem(args, 1);
+   Py_INCREF(pyfunc);
+
+   if (!PyObject_HasAttrString(instance, "input_types")) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: No input_types attribute found.");
+      return NULL;
+   }
+
+   if (!PyObject_HasAttrString(instance, "return_type")) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: No return_type attribute found.");
+      return NULL;
+   }
+
+   auto inputTypes = PyObject_GetAttrString(instance, "input_types");
+   auto returnType = PyObject_GetAttrString(instance, "return_type");
+
+   // Extract name of Python callable
+   PyObject* pyname;
+   if (PyObject_HasAttrString(instance, "name")) {
+      pyname = PyObject_GetAttrString(instance, "name");
+   } else {
+      if (!PyObject_HasAttrString(pyfunc, "__name__")) {
+         PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: Python callable does not have attribute __name__.");
+         return NULL;
+      }
+      pyname = PyObject_GetAttrString(pyfunc, "__name__");
+   }
+
+   auto name = CPyCppyy_PyUnicode_AsString(pyname);
+   Py_DECREF(pyname);
+
+   // Get C++ return type
+   if (!CPyCppyy_PyUnicode_Check(returnType)) {
+      PyErr_SetString(PyExc_RuntimeError, "Failed to create C++ callable: Return type argument cannot be interpreted as string.");
+      return NULL;
+   }
+   std::string returnTypeStr = CPyCppyy_PyUnicode_AsString(returnType);
+   Py_DECREF(returnType);
+   if (returnTypeStr.compare("") == 0) {
+      returnTypeStr = "void";
+   }
+
+   // Find numba types for C++ types
+   std::map<std::string, std::string> typemap = {
```

Comment:
This would be different between numba and the generic case. I would try to have helper functions whose behaviour is different depending on if it is numba or generic, and other functions that are common to both.